### PR TITLE
Allow force closing in any state

### LINKED ISSUE #113: LN-DLC: Allow force-close at any point
We must let consumers trigger force-close at any point in the protocol.

I think we will need to handle some states differently. For instance, with an `Offered` state I think we would need to fall back to the previous commitment transaction (sans split transaction), whereas with `CloseConfirmed` we would need to use the newly negotiated state (sans split transaction too).

I open this issue because of this check which ensures that the subchannel is in the `Signed` state:

https://github.com/p2pderivatives/rust-dlc/blob/add0b0a7f1ade21a31bad2c6781fe6864980ce25/dlc-manager/src/sub_channel_manager.rs#L709-L714

### PR DESCRIPTION
**Pull request by Tibo-lg**
This PR makes it possible to force close a sub channel in (almost) every states.

Closes #113

### CODE REVIEW

File: dlc-manager/src/sub_channel_manager.rs
Comment (luckysori) [inline dlc-manager/src/sub_channel_manager.rs:1055]: I think we should document that this MUST be called in the LDK event handler.

File: dlc-manager/src/sub_channel_manager.rs
Comment (luckysori) [inline dlc-manager/src/sub_channel_manager.rs:1082]: It might be cleaner to do this in all cases after the match. We could `if let Some` the `dlc_channel_id`.

**Review by luckysori — APPROVED** [review]
Nice job! The solution turned out to be not so complicated. I left questions and a few comments hoping to make things a bit clearer. Nothing blocking AFAICT.

Obviously the tests are getting quite messy, but we can tackle that separately.

File: dlc-manager/src/subchannel/mod.rs
Comment (Tibo-lg) [reply dlc-manager/src/subchannel/mod.rs:1]: Yes you can there are even tests for it, what was your concern?

Comment (luckysori) [reply dlc-manager/src/subchannel/mod.rs:1]: I left this comment early in the review and forgot to delete it. Sorry! I was confused because you didn't need to modify the state for it to work, unlike the other states.

File: dlc-manager/src/channel/mod.rs
Comment (Tibo-lg) [reply dlc-manager/src/channel/mod.rs]: Thanks fixed

File: dlc-manager/src/channel/mod.rs
Comment (Tibo-lg) [reply dlc-manager/src/channel/mod.rs]: Sure updated

File: dlc-manager/src/subchannel/mod.rs
Comment (Tibo-lg) [reply dlc-manager/src/subchannel/mod.rs]: Updated

File: dlc-manager/src/subchannel/mod.rs
Comment (Tibo-lg) [reply dlc-manager/src/subchannel/mod.rs]: I expanded the comments a bit, let me know if it's easier to understand!

File: dlc-manager/tests/channel_execution_tests.rs
Comment (Tibo-lg) [reply dlc-manager/tests/channel_execution_tests.rs:1415]: Yes thanks (remains from some debugging...)

File: dlc-manager/tests/ln_dlc_channel_execution_tests.rs
Comment (Tibo-lg) [reply dlc-manager/tests/ln_dlc_channel_execution_tests.rs:406]: Yes, I think at some point it might be better to have the LDK `ChannelManager` built within the `SubChannelManager` so that we can make sure that everything is properly setup, but I think this PR is already big enough :)

File: dlc-manager/tests/ln_dlc_channel_execution_tests.rs
Comment (Tibo-lg) [reply dlc-manager/tests/ln_dlc_channel_execution_tests.rs:740]: No, once the `CloseFinalize` is processed, the node is in `OffChainClosed` state, which should be handled through LDK (at least that's how it's currently implemented).

File: dlc-manager/tests/ln_dlc_channel_execution_tests.rs
Comment (Tibo-lg) [reply dlc-manager/tests/ln_dlc_channel_execution_tests.rs:173]: Added comments, I hope it's clearer!

File: dlc-manager/tests/ln_dlc_channel_execution_tests.rs
Comment (Tibo-lg) [reply dlc-manager/tests/ln_dlc_channel_execution_tests.rs:1]: Yeah sorry...

File: dlc-manager/tests/ln_dlc_channel_execution_tests.rs
Comment (luckysori) [reply dlc-manager/tests/ln_dlc_channel_execution_tests.rs:406]: Yeah, that sounds like a pretty big change!

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:676]: Is it enough with the comments now or you still would prefer to rename? I don't have a strong opinion but using a long descriptive name can sometimes hinder readability of the code.

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs]: Put comments, let me know what you think!

File: dlc-manager/tests/ln_dlc_channel_execution_tests.rs
Comment (luckysori) [reply dlc-manager/tests/ln_dlc_channel_execution_tests.rs:1]: I'm just mentioning it in case you _want_ me to take a closer look before merging 😁

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:737]: Yeah it's quite different compared to #120. I'm not saying it's bulletproof, but it should be safe to rollback the state here. At worst, the closing fails, you're out of sync with your peer, but since you're trying to force close anyway it shouldn't matter I think.

File: dlc-manager/src/sub_channel_manager.rs
Comment (luckysori) [reply dlc-manager/src/sub_channel_manager.rs:676]: Yep, comments are good!

I agree that long names aren't always good (I usually go for long names haha).

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:759]: Mmmm I feel maybe I should give up this rollback. It's there mainly because I was lazy to handle too many states and wanted to reuse code, but I'm thinking it might not have been a good idea.

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs]: updated to `force_close_with_ldk`, hope that works for you!

File: dlc-manager/src/sub_channel_manager.rs
Comment (luckysori) [reply dlc-manager/src/sub_channel_manager.rs:737]: Right, the real problem about rolling back is to agree with your counterparty on doing so.

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs]: I put quite some docs within the `force_close_sub_channel` method, is it enough or do you still want more here?

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:700]: This is required for the rollback of the state in `CloseOffered` and `CloseAccepted`.

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs]: Done

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs]: Done

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs]: We can not cleanup, but it means storage will just keep growing as channels get open/closed. It's probably not a big issue and I'm fine either way. But if we do want to cleanup, it's required here because our chain monitor will not react to commitment transactions.

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs]: Thanks fixed!

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:820]: Commented let me know.

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:842]: Indeed, not much at least

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs]: Fixed

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:1070]: No this is to react to the counter party force closing the channel using a transaction that we don't track (e.g. a Lightning commitment transaction directly attached to the funding output).

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs]: Done

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:1096]: Mmm I guess for now it'd be ok to do it, but ideally I think it should be differentiated because as I put in comments it's not always safe to cleanup right away.

File: dlc-manager/tests/ln_dlc_channel_execution_tests.rs
Comment (Tibo-lg) [reply dlc-manager/tests/ln_dlc_channel_execution_tests.rs:1]: I'm rather confident that it covers the cases that I wanted to since I had to debug for pretty much each of them :D . But of course if you have extra motivation it's always nice to have a second pair of eyes checking things :)

File: dlc-manager/src/sub_channel_manager.rs
Comment (luckysori) [reply dlc-manager/src/sub_channel_manager.rs:1070]: Interesting. How is that different to LDK closing the channel automatically though? I could imagine us using this method to handle that too.

Obviously we don't have to do it now, but that's what I first assumed you were doing when looking at this method.

Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:1070]: Yes it will be used for the automatic close as well, just not yet. That's why e.g. the `Signed` state error says that we don't handle it yet.

Comment (luckysori) [reply dlc-manager/src/sub_channel_manager.rs:1070]: Perfect, makes sense!

File: dlc-manager/src/sub_channel_manager.rs
Comment (Tibo-lg) [reply dlc-manager/src/sub_channel_manager.rs:759]: Actually I realized this rollback will not be persisted if the force closing fail, so I think it's not a big deal at the end.

**Review by luckysori — APPROVED** [review]
LGTM, I think we can merge!

**Pull request merged**
